dask_kubernetes/core.py

```diff
@@ -14,19 +17,186 @@
     yaml = False
 
 import dask
-from distributed.deploy import LocalCluster, Cluster
+from distributed.deploy import SpecCluster, ProcessInterface
 from distributed.comm.utils import offload
-import kubernetes
+from distributed.utils import Log, Logs
+import kubernetes_asyncio as kubernetes
 from tornado import gen
 
-from .objects import make_pod_from_dict, clean_pod_template
+from .objects import (
+    make_pod_from_dict,
+    make_service_from_dict,
+    clean_pod_template,
+    clean_service_template,
+)
 from .auth import ClusterAuth
-from .logs import Log, Logs
 
 logger = logging.getLogger(__name__)
 
 
-class KubeCluster(Cluster):
+class Pod(ProcessInterface):
+    """ A superclass for Kubernetes Pods
+    See Also
+    --------
+    Worker
+    Scheduler
+    """
+
+    def __init__(self, core_api, pod_template, namespace, loop=None, **kwargs):
+        self.pod = None
+        self.core_api = core_api
+        self.pod_template = copy.deepcopy(pod_template)
+        self.namespace = namespace
+        self.name = None
+        self.loop = loop
+        self.kwargs = kwargs
+        super().__init__()
+
+    @property
+    def cluster_name(self):
+        return self.pod_template.metadata.labels["dask.org/cluster-name"]
+
+    async def start(self, **kwargs):
+        self.pod = await self.core_api.create_namespaced_pod(
+            self.namespace, self.pod_template
+        )
+
+        while self.pod.status.phase == "Pending":
+            self.pod = await self.core_api.read_namespaced_pod(
+                self.pod.metadata.name, self.namespace
+            )
+            await asyncio.sleep(0.1)
+
+        while self.address is None:
+            logs = await self.logs()
+            for line in logs.splitlines():
+                for query_string in ["worker at:", "Scheduler at:"]:
+                    if query_string in line:
+                        self.address = line.split(query_string)[1].strip()
+            await asyncio.sleep(0.1)
```

Comment:
In other SpecCluster we treat a worker Job as "started" once it has been successfully handed off to the resource manager.  We don't actually wait for it to switch from a pending to an active state.  This might not be correct, but currently I think that various other systems, like adaptive, expect this sort of behavior.  The Cluster class will separately poll the scheduler to find out who has arrived.  

Note that this isn't true for the scheduler, for whom we do actually need the address.

Reply:
That sounds reasonable. We must poll for the address before leaving the function here though. Do you propose I call the super start (which switches the status) higher up?

Reply:
We might have to treat schedulers differently than workers in this respect.  I don't think we need the worker address any more.

Reply:
Ah ok!